### ISSUE TITLE
Basic type checking (possibly enhanced by using clojure.spec specs)

### ISSUE THREAD

**Issue by borkdude** [borkdude/clj-kondo#23]
When reading specs like

`(s/fdef foo :args (s/cat :i int?))`

we can detect wrong calls like `(foo "string")`.

We could also offer a suggestion when the user types `(foo _)`.

We might use [speculative](https://github.com/borkdude/speculative) for checking core calls.

**Comment by kennyjwilli** [comment]
This is a great idea! May be worth checking out [spectrum](https://github.com/arohner/spectrum).

**Comment by borkdude** [comment]
Note to self:

In case I want to use spec itself with this inside the binary, this is a working example with GraalVM:

https://twitter.com/squeekeeper/status/1115531608835198976

**Comment by kennyjwilli** [comment]
Another example: 

We often have functions that take their arguments in a map. Being able to statically determine if you are calling a function with an incomplete arg-map would be amazing. 
```clojure
(defn foo 
  [{:keys [a b]}]
  (+ a b))

(s/def ::a int?)
(s/def ::b int?)

(s/fdef foo 
  :args (s/cat :arg-map (s/keys :req-un [::a ::b]))
  :ret int?)

;; mark that foo should include the :b key and that :a should be an int?
(foo {:a "wrong"})
```

**Comment by borkdude** [comment]
A basic for of type checking is implemented in #445. Spec enhancements to do this may be possible, not sure if and when.
